File: Baekjun/10.1/2228. 구간 나누기/solution.swift
let NM = readLine()!.split(separator: " ").compactMap { Int($0) }
let N = NM[0], M = NM[1]

let MIN = -1 * 3276800

// var con = Array(repeating: [0]+Array(repeating: MIN, count: M), count: N+1)
// var noncon = Array(repeating: [0]+Array(repeating: MIN, count: M), count: N+1)
//
// for i in 1 ... N {
//    let num = Int(readLine()!)!
//    for j in 1 ... min(M, (i+1) / 2) {
//        con[i][j] = max(con[i-1][j], noncon[i-1][j-1])+num
//        noncon[i][j] = max(con[i-1][j], noncon[i-1][j])
//    }
// }
//
// print(max(con.last!.last!, noncon.last!.last!))

var dp = Array(repeating: [0] + Array(repeating: MIN, count: M), count: N + 1)

var prefix: [Int] = []

for _ in 0 ..< N {
    let num = Int(readLine()!)!
    if prefix.isEmpty {
        prefix.append(num)
    } else {
        prefix.append(prefix.last! + num)
    }
}

for i in 1 ..< N + 1 {
    for j in 1 ..< M + 1 {
        dp[i][j] = dp[i-1][j]
        for k in 1 ... i {
            if k >= 2 {
                dp[i][j] = max(dp[i][j], dp[k-2][j-1] + prefix[i-1]-prefix[k-2])
            } else if k == 1, j == 1 {
                dp[i][j] = max(dp[i][j], prefix[i-1])
            }
        }
    }
}

print(dp.last!.last!)

File: Baekjun/10.1/2228. 구간 나누기/solution.py
N, M = map(int, input().split())

MIN = -3276800

# dp = [[0] + [MIN] * (M) for _ in range(N + 1)]

# total = 0
# prefix = [0]

# for _ in range(N):
#     total += int(input())
#     prefix.append(total)

# for i in range(1, N + 1):
#     for j in range(1, M + 1):
#         dp[i][j] = dp[i - 1][j]
#         for k in range(1, i + 1):
#             if k >= 2:
#                 dp[i][j] = max(dp[i][j], dp[k - 2][j - 1] + prefix[i] - prefix[k - 1])
#             elif k == 1 and j == 1:
#                 dp[i][j] = max(dp[i][j], prefix[i])
# print(dp[N][M])

con = [[0] + [MIN] * M for _ in range(N + 1)]
noncon = [[0] + [MIN] * M for _ in range(N + 1)]

for i in range(1, N + 1):
    n = int(input())
    for j in range(1, min(M, (i + 1) // 2) + 1):
        noncon[i][j] = max(noncon[i - 1][j], con[i - 1][j])
        con[i][j] = max(con[i - 1][j], noncon[i - 1][j - 1]) + n

print(max(con[-1][-1], noncon[-1][-1]))
